monitoring/monitoring.go

```diff
@@ -0,0 +1,633 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package monitoring AOS Core Monitoring Component
+package monitoring
+
+import (
+	"container/list"
+	"context"
+	"math"
+	"runtime"
+	"sync"
+	"time"
+
+	"github.com/aoscloud/aos_common/aoserrors"
+	"github.com/aoscloud/aos_common/aostypes"
+	"github.com/aoscloud/aos_common/api/cloudprotocol"
+	"github.com/aoscloud/aos_common/utils/fs"
+	"github.com/shirou/gopsutil/cpu"
+	"github.com/shirou/gopsutil/disk"
+	"github.com/shirou/gopsutil/mem"
+	"github.com/shirou/gopsutil/process"
+	log "github.com/sirupsen/logrus"
+)
+
+/***********************************************************************************************************************
+ * Consts
+ **********************************************************************************************************************/
+
+// Service status.
+const (
+	MinutePeriod = iota
+	HourPeriod
+	DayPeriod
+	MonthPeriod
+	YearPeriod
+)
+
+// For optimization capacity should be equals numbers of measurement values
+// 5 - RAM, CPU, UsedDisk, InTraffic, OutTraffic.
+const capacityAlertProcessorElements = 5
+
+/***********************************************************************************************************************
+ * Types
+ **********************************************************************************************************************/
+
+// AlertSender interface to send resource alerts.
+type AlertSender interface {
+	SendAlert(alert cloudprotocol.AlertItem)
+}
+
+// Sender sends alerts to the cloud.
+type MonitoringSender interface {
+	SendMonitoringData(monitoringData cloudprotocol.MonitoringData) (err error)
+}
+
+// TrafficMonitoring interface to get network traffic.
+type TrafficMonitoring interface {
+	GetSystemTraffic() (inputTraffic, outputTraffic uint64, err error)
+	GetInstanceTraffic(instanceID string) (inputTraffic, outputTraffic uint64, err error)
+}
+
+// Monitoring configuration for system monitoring.
+type Monitoring struct {
+	SendPeriod aostypes.Duration `json:"sendPeriod"`
+	PollPeriod aostypes.Duration `json:"pollPeriod"`
+	aostypes.ServiceAlertRules
+}
+
+// Monitor instance.
+type Monitor struct {
+	sync.Mutex
+
+	alertSender      AlertSender
+	monitoringSender MonitoringSender
+
+	config MonitorConfigParams
+
+	sendTimer *time.Ticker
+	pollTimer *time.Ticker
+
+	globalMonitoringData cloudprotocol.GlobalMonitoringData
+
+	alertProcessors *list.List
+
+	instanceMonitoringMap map[string]*instanceMonitoring
+	trafficMonitoring     TrafficMonitoring
+
+	cancelFunction context.CancelFunc
+}
+
+// MonitorParams instance monitoring parameters.
+type MonitorParams struct {
+	cloudprotocol.InstanceIdent
+	UID        int
+	GID        int
+	AlertRules *aostypes.ServiceAlertRules
+}
+
+// MonitorConfigParams instance monitoring config parameters.
+type MonitorConfigParams struct {
```

Comment:
move this struct closed to Monitoring struct, just t increase code readability

Reply:
Done